### ISSUE TITLE
Страшная ошибка при циклических записях путём шаблонов

### ISSUE THREAD

**Issue by SunSerega** [pascalabcnet/pascalabcnet#2799]
```
type
  r0<T> = record
    val: T;
  end;
  
  r1 = record
    a: r0<r1>;
    // То же самое, но с Nullable вместо r0
//    a: r1?;
  end;
  
begin
  Println(new r1);
end.
```
```
Ошибка времени выполнения: System.TypeLoadException: Не удалось загрузить тип "2.r1" из сборки "2, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null".
Стек:
   в 2.Program.$Main()
   в 2.Program.Main()
```
Ошибка не указывает что проблема в циклическом поле `r1.a`.
Для тех кто не знает как работает `a: r1?` - может быть сложно понять, что тут пошло не так.

**Comment by samuraiGH** [comment]
А разве можно повлиять на ошибки, которые кидает CLR?

**Comment by SunSerega** [comment]
Ошибка об этом должна выдаваться ещё во время компиляции, как и для этого случая:
```
  r1 = record
    a: r1;
  end;
```
Сделать рекурсивную проверку для этого тривиально - похожих уже полно в компиляторе.

А до среды выполнения сломанные типы (`TypeLoadException`) и тела методов (`InvalidProgramException`: "Среда выполнения обнаружила недопустимую программу"), доходить не должны - это всё issue именно компилятора.

**Comment by miks1965** [comment]
По идее надо проверять что поле является структурой и шаблонный параметр её - тоже структура, которая даёт цикл

**Comment by SunSerega** [comment]
Нет, это не достаточная причина запрещать - в `r0<T>` может не быть поля типа `T`.
Вместо этого надо сделать рекурсивную функцию. В случае `a: r0<r1>` она должна проверять, чтобы в `r0<r1>` не было полей типов `r1` и `r0<r1>`.
При чём проверка на равенство типов нужна строгая, потому что:
```
  r1 = record
    a: r0<r0<byte>>;
  end;
```
Это вполне валидный код и сейчас он работает без проблем.
`r0<r0<byte>>` содержит поле типа `r0<byte>` - это нормально.

**Comment by miks1965** [comment]
Я имею в виду, что поле записи r1 - это запись, параметризованная типом записи, которая включает r1 прямо или косвенно.
